File: web/src/lib/operations/recommendations.ts
import type {
  ShopProfileType,
  ValidatedWorkflowId,
} from "@/lib/mock-data/operations/schemas";

import type { ShopProfileClassification } from "./classification";
import type { HealthCheckResults, HealthIndicatorSeverity } from "./health-check";
import { REFUND_SPIKE_THRESHOLD_RATIO } from "./health-check";
import { getWorkflowsForProfile } from "./workflow-catalog";

/**
 * Product-owned MID_LARGE_SHOP impact filter (VND). Null until recorded in EXECUTION.md.
 * Mock P1.8 ranks all eligible workflows when unset (ADR-026).
 */
export const MID_LARGE_IMPACT_THRESHOLD_VND: number | null = null;

export type ImpactConfidence = "high" | "medium" | "low";

export interface WorkflowExpectedImpact {
  metric: string;
  value: number;
  confidence: ImpactConfidence;
}

export interface WorkflowRecommendation {
  workflow_id: ValidatedWorkflowId;
  workflow_name: string;
  priority: number;
  rationale: string;
  expected_impact: WorkflowExpectedImpact;
  preconditions_met: boolean;
  user_action_required: boolean;
}

/**
 * Stable P1.8 output envelope for the operations pipeline ranking stage.
 * P2 swaps scoring; schema shape remains stable per ADR-026.
 */
export interface WorkflowRecommendations {
  shop_profile: ShopProfileType;
  recommended_workflows: WorkflowRecommendation[];
}

interface ScoredWorkflow {
  workflowId: ValidatedWorkflowId;
  score: number;
  rationale: string;
  expectedImpact: WorkflowExpectedImpact;
  preconditionsMet: boolean;
}

const WORKFLOW_DISPLAY_NAMES: Record<ValidatedWorkflowId, string> = {
  npl: "Thêm sản phẩm mới",
  minimize_violations: "Giảm thiểu vi phạm",
  budget_optimization: "Tối ưu ngân sách quảng cáo",
  product_scaling: "Mở rộng sản phẩm",
  refund_spike_detection: "Phát hiện đỉnh hoàn tiền",
  stockout_prevention: "Phòng tránh hết hàng",
};

function severityToConfidence(severity: HealthIndicatorSeverity): ImpactConfidence {
  if (severity === "critical") return "high";
  if (severity === "warning") return "medium";
  return "low";
}

function scoreNewShopWorkflow(
  workflowId: ValidatedWorkflowId,
  health: HealthCheckResults,
): ScoredWorkflow {
  const probation = health.indicators.probation_progress;
  const sps = health.indicators.sps_health;
  const ahr = health.indicators.ahr_health;

  if (workflowId === "npl") {
    const urgency =
      100 -
      probation.percent_toward_graduation +
      sps.threshold_gap * 10 +
      Math.max(0, 30 - probation.days_remaining);

    return {
      workflowId,
      score: urgency,
      rationale: probation.summary || sps.summary,
      expectedImpact: {
        metric: "SPS",
        value: Math.round(sps.threshold_gap * 10) / 10,
        confidence: severityToConfidence(sps.severity),
      },
      preconditionsMet: probation.severity !== "not_applicable" && sps.sps_current > 0,
    };
  }

  const violationWeight = ahr.severity === "critical" ? 40 : 20;
  const urgency = ahr.threshold_gap * 2 + violationWeight;

  return {
    workflowId,
    score: urgency,
    rationale: ahr.summary,
    expectedImpact: {
      metric: "AHR",
      value: ahr.threshold_gap,
      confidence: severityToConfidence(ahr.severity),
    },
    preconditionsMet: ahr.ahr_current > 0 && ahr.threshold_gap > 0,
  };
}

function scoreMidLargeWorkflow(
  workflowId: ValidatedWorkflowId,
  health: HealthCheckResults,
): ScoredWorkflow {
  switch (workflowId) {
    case "budget_optimization": {
      const indicator = health.indicators.ad_roas_efficiency;
      const score =
        indicator.active_campaigns_below_target_pct +
        (indicator.severity === "critical" ? 30 : indicator.severity === "warning" ? 15 : 0);

      return {
        workflowId,
        score,
        rationale: indicator.summary,
        expectedImpact: {
          metric: "ROAS",
          value: indicator.active_campaigns_below_target_pct,
          confidence: severityToConfidence(indicator.severity),
        },
        preconditionsMet:
          indicator.severity !== "not_applicable" && indicator.campaigns.length >= 2,
      };
    }
    case "product_scaling": {
      const indicator = health.indicators.product_scaling_opportunity;
      const topScore = indicator.top_skus[0]?.growth_potential_score ?? 0;

      return {
        workflowId,
        score: topScore,
        rationale: indicator.summary,
        expectedImpact: {
          metric: "doanh_thu_SKU",
          value: topScore,
          confidence: severityToConfidence(indicator.severity),
        },
        preconditionsMet:
          indicator.severity !== "not_applicable" && indicator.top_skus.length > 0,
      };
    }
    case "refund_spike_detection": {
      const indicator = health.indicators.refund_spike_indicator;
      const score = indicator.spike_detected
        ? 100 + indicator.percent_change_vs_baseline
        : indicator.percent_change_vs_baseline;

      return {
        workflowId,
        score,
        rationale: indicator.summary,
        expectedImpact: {
          metric: "tỷ_lệ_hoàn",
          value: Math.round(indicator.refund_rate_7d * 1000) / 10,
          confidence: severityToConfidence(indicator.severity),
        },
        preconditionsMet:
          indicator.spike_detected ||
          indicator.percent_change_vs_baseline / 100 > REFUND_SPIKE_THRESHOLD_RATIO,
      };
    }
    case "stockout_prevention": {
      const indicator = health.indicators.inventory_health;
      const score =
        indicator.at_risk_sku_count * 25 +
        (indicator.severity === "critical" ? 20 : indicator.severity === "warning" ? 10 : 0);

      return {
        workflowId,
        score,
        rationale: indicator.summary,
        expectedImpact: {
          metric: "SKU_rủi_ro",
          value: indicator.at_risk_sku_count,
          confidence: severityToConfidence(indicator.severity),
        },
        preconditionsMet:
          indicator.severity !== "not_applicable" && indicator.skus.length > 0,
      };
    }
    default:
      return {
        workflowId,
        score: 0,
        rationale: "",
        expectedImpact: { metric: "unknown", value: 0, confidence: "low" },
        preconditionsMet: false,
      };
  }
}

function passesImpactThreshold(expectedImpactValue: number): boolean {
  if (MID_LARGE_IMPACT_THRESHOLD_VND === null) {
    return true;
  }

  return expectedImpactValue >= MID_LARGE_IMPACT_THRESHOLD_VND;
}

function toRecommendation(item: ScoredWorkflow, priority: number): WorkflowRecommendation {
  return {
    workflow_id: item.workflowId,
    workflow_name: WORKFLOW_DISPLAY_NAMES[item.workflowId],
    priority,
    rationale: item.rationale,
    expected_impact: item.expectedImpact,
    preconditions_met: item.preconditionsMet,
    user_action_required: item.preconditionsMet,
  };
}

/**
 * Rules-based workflow ranking (P1.8-4).
 * Profile-gates to WORKFLOW_CATALOG; never surfaces IDs outside the validated catalog.
 */
export function rankWorkflowRecommendations(
  profile: ShopProfileClassification,
  health: HealthCheckResults,
): WorkflowRecommendations {
  const eligibleWorkflows = getWorkflowsForProfile(profile);

  const scored = eligibleWorkflows.map((workflowId) =>
    profile === "NEW_SHOP"
      ? scoreNewShopWorkflow(workflowId, health)
      : scoreMidLargeWorkflow(workflowId, health),
  );

  const filtered =
    profile === "MID_LARGE_SHOP"
      ? scored.filter((item) => passesImpactThreshold(item.expectedImpact.value))
      : scored;

  const ranked = [...filtered].sort((left, right) => right.score - left.score);

  return {
    shop_profile: profile,
    recommended_workflows: ranked.map((item, index) => toRecommendation(item, index + 1)),
  };
}

File: backend/src/juli_backend/services/scoring/recommendations.py
"""Rank execution_layer workflows from visual_layer advisory signals (#303)."""

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass

from juli_backend.services.aggregates.types import ShopProfile
from juli_backend.services.scoring.kpi_catalog import (
    WORKFLOW_DISPLAY_NAMES,
    get_workflows_for_profile,
)
from juli_backend.services.scoring.types import (
    AdvisorySignal,
    ImpactConfidence,
    ScoringSignals,
    WorkflowExpectedImpact,
    WorkflowRecommendation,
    WorkflowRecommendations,
)

_SEVERITY_SCORE = {
    "critical": 100.0,
    "warning": 50.0,
    "healthy": 10.0,
    "not_applicable": 0.0,
}


@dataclass(frozen=True)
class _ScoredWorkflow:
    workflow_key: str
    score: float
    rationale: str
    expected_impact: WorkflowExpectedImpact
    preconditions_met: bool
    source_kpi_ids: tuple[str, ...]


def _severity_to_confidence(severity: str) -> ImpactConfidence:
    if severity == "critical":
        return "high"
    if severity == "warning":
        return "medium"
    return "low"


def _impact_value(signal: AdvisorySignal) -> float:
    if signal.kpi_id == "return_request_rate" and signal.signal_type == "risk":
        return signal.severity == "critical" and 100.0 or 50.0
    return _SEVERITY_SCORE[signal.severity]


def _score_workflows(
    profile: ShopProfile,
    signals: ScoringSignals,
) -> list[_ScoredWorkflow]:
    eligible = get_workflows_for_profile(profile)
    by_workflow: dict[str, list[AdvisorySignal]] = defaultdict(list)

    for signal in signals.kpis.values():
        if signal.signal_type == "unavailable":
            continue
        for key in signal.workflow_keys:
            if key in eligible:
                by_workflow[key].append(signal)

    scored: list[_ScoredWorkflow] = []
    for workflow_key, linked in by_workflow.items():
        best = max(
            linked,
            key=lambda item: (_SEVERITY_SCORE[item.severity], item.signal_type == "risk"),
        )
        score = _SEVERITY_SCORE[best.severity] + (10.0 if best.signal_type == "risk" else 0.0)
        source_kpis = tuple(sorted({item.kpi_id for item in linked}))
        scored.append(
            _ScoredWorkflow(
                workflow_key=workflow_key,
                score=score,
                rationale=best.one_line,
                expected_impact=WorkflowExpectedImpact(
                    metric=best.kpi_id,
                    value=_impact_value(best),
                    confidence=_severity_to_confidence(best.severity),
                ),
                preconditions_met=best.severity in {"critical", "warning", "healthy"},
                source_kpi_ids=source_kpis,
            )
        )

    return sorted(scored, key=lambda item: item.score, reverse=True)


def rank_workflow_recommendations(
    profile: ShopProfile,
    signals: ScoringSignals,
) -> WorkflowRecommendations:
    ranked = _score_workflows(profile, signals)
    return WorkflowRecommendations(
        shop_profile=profile,
        recommended_workflows=[
            WorkflowRecommendation(
                workflow_key=item.workflow_key,
                workflow_name=WORKFLOW_DISPLAY_NAMES.get(
                    item.workflow_key, item.workflow_key
                ),
                priority=index + 1,
                rationale=item.rationale,
                expected_impact=item.expected_impact,
                preconditions_met=item.preconditions_met,
                user_action_required=item.preconditions_met,
                source_kpi_ids=item.source_kpi_ids,
            )
            for index, item in enumerate(ranked)
        ],
    )
